activerecord/lib/active_record/insert_all.rb

```diff
@@ -15,33 +15,49 @@ def execute(relation, ...)
       end
     end
 
-    def initialize(relation, connection, inserts, on_duplicate:, update_only: nil, returning: nil, unique_by: nil, record_timestamps: nil)
+    def column_mode?
+      @column_mode
+    end
+
+    def initialize(relation, connection, inserts, on_duplicate:, update_only: nil, returning: nil, unique_by: nil, record_timestamps: nil, columns: nil)
       @relation = relation
-      @model, @connection, @inserts = relation.model, connection, inserts.map(&:stringify_keys)
+      @model, @connection = relation.model, connection
+      @column_mode = !!columns
+      if @column_mode
+        @keys = columns.map(&:to_s).to_set
```

Comment:
What happens if some column names were duplicated?

Reply:
Good call -- I'll need to guard against that.